Win32.xs

```diff
@@ -1682,6 +1693,248 @@ XS(w32_IsDeveloperModeEnabled)
+     * target server (using the default WINHTTP_ACCESS_TYPE_NO_PROXY
+     * configuration, which the request handle will inherit from the session).
+     */
+    if (hRequest) {
+        ZeroMemory(&AutoProxyOptions, sizeof(AutoProxyOptions));
+        ZeroMemory(&ProxyInfo, sizeof(ProxyInfo));
+        AutoProxyOptions.dwFlags = WINHTTP_AUTOPROXY_AUTO_DETECT;
+        AutoProxyOptions.dwAutoDetectFlags =
+                                    WINHTTP_AUTO_DETECT_TYPE_DHCP |
+                                    WINHTTP_AUTO_DETECT_TYPE_DNS_A;
+        AutoProxyOptions.fAutoLogonIfChallenged = TRUE;
+
+        if(WinHttpGetProxyForUrl(hSession,
+                                url,
+                                &AutoProxyOptions,
+                                &ProxyInfo)) {
+            if(!WinHttpSetOption(hRequest,
+                                WINHTTP_OPTION_PROXY,
+                                &ProxyInfo,
+                                cbProxyInfoSize)) {
+                bAborted = TRUE;
+                Perl_warn(aTHX_ "Win32::HttpGetFile: setting proxy options failed");
+            }
+            Safefree(ProxyInfo.lpszProxy);
+            Safefree(ProxyInfo.lpszProxyBypass);
+        }
+    }
+
+    /* Send a request. */
+    if (hRequest && !bAborted)
+        bResults = WinHttpSendRequest(hRequest,
+                                      WINHTTP_NO_ADDITIONAL_HEADERS,
+                                      0,
+                                      WINHTTP_NO_REQUEST_DATA,
+                                      0,
+                                      0,
+                                      0);
+
+    /* End the request. */
+    if (bResults)
+        bResults = WinHttpReceiveResponse(hRequest, NULL);
+
+    /* Create output file for download. */
+    if (bResults) {
+        hOut = CreateFileW(file,
+                           GENERIC_WRITE,
+                           FILE_SHARE_READ | FILE_SHARE_WRITE,
+                           NULL,
+                           CREATE_ALWAYS,
+                           FILE_ATTRIBUTE_NORMAL,
+                           NULL);
+
+        if (!hOut || hOut == INVALID_HANDLE_VALUE)
+            bFileError = TRUE;
+    }
+
+    /* Keep checking for data until there is nothing left. */
+    if (!bFileError && bResults) {
+        do {
+            /* Check for available data. */
+            dwSize = 0;
+            if (!WinHttpQueryDataAvailable(hRequest, &dwSize)) {
+                bAborted = TRUE;
+                break;
+            }
+
+            /* No more available data. */
+            if (!dwSize)
+                break;
+
+            /* Allocate space for the buffer. */
+            New(0, pszOutBuffer, dwSize + 1, char);
+            if (!pszOutBuffer) {
+                bAborted = TRUE;
+                break;
+            }
+
+            /* Read the Data. */
+            ZeroMemory(pszOutBuffer, dwSize+1);
```

Comment:
Zeroing isn't necessary, WinHttpReadData returns the length of the data.

Reply:
> Zeroing isn't necessary, WinHttpReadData returns the length of the data.

It might be paranoia, but it doesn't really hurt anything.  Like most of the code here, I lifted it straight from the examples in the docs, so at least somebody at Microsoft considers it a best practice.  In the context of downloading a file, a little zeroing of memory that's not strictly necessary is probably in the noise performance-wise.

There are various performance optimizations that could be considered, such as buffering up multiple reads from the network before doing any file I/O to avoid doing lots of small writes.  My goal for the moment is to get something reliably working that is as simple as possible.  If performance is problematic, further work can always be done to address that.

Reply:
>somebody at Microsoft considers it a best practice

They zero the buffer to ensure it's null-terminated so it can be passed to printf(). 

> My goal for the moment is to get something reliably working that is as simple as possible. If performance is problematic, further work can always be done to address that.

It isn't just for performance reasons. Removing zeroing makes it a bit simpler.